tokio/src/sync/notify.rs

```diff
@@ -516,12 +529,23 @@ impl Notify {
             return;
         }
 
-        // At this point, it is guaranteed that the state will not
-        // concurrently change, as holding the lock is required to
-        // transition **out** of `WAITING`.
+        // Increment the number of times this method was called
+        // and transition to empty.
+        let new_state = set_state(inc_num_notify_waiters_calls(curr), EMPTY);
+        self.state.store(new_state, SeqCst);
+
+        let decoupled_list = std::mem::take(&mut *waiters);
+
+        let guard = UnsafeCell::new(Waiter::new());
```

Comment:
We probably want to pin `guard` to ensure it doesn't accidentally move (the `pin!` macro might work here). Also, could you add a *big* comment saying it is critical for safety that `guard` does not move and is not dropped until the guarded list is dropped?